[lldb] Synchronize the debugger's stdout and stderr streams

JDevlieghere opened the pull request:
This patch improves the synchronization of the debugger's output and
error streams using two new abstractions: LockableStreamFile and
LockedStreamFile.

 - LockableStreamFile is a wrapper around a StreamFile and a mutex.
   Client cannot use the StreamFile without calling Lock, which returns
   a LockedStreamFile.

 - LockedStreamFile is an RAII object that locks the stream for the
   duration of its existence.  As long as you hold on to the returned
   object you are permitted to write to the stream. The destruction of
   the object automatically flush the output stream.

bulbazord commented on lldb/source/Core/IOHandlerCursesGUI.cpp:
```diff
@@ -7574,7 +7574,9 @@ IOHandlerCursesGUI::IOHandlerCursesGUI(Debugger &debugger)
   IOHandler::Activate();
   if (!m_app_ap) {
-    m_app_ap = std::make_unique<Application>(GetInputFILE(), GetOutputFILE());
+    m_app_ap = std::make_unique<Application>(
```
nit: while you're here, `m_app_ap` -> `m_app_up`? auto pointers are dead, long live auto pointers!

JDevlieghere replied:
Addressed in 776fa2d731c17d6ba0afad2554ebc89cf5e3e5ef

bulbazord commented on lldb/source/Host/common/Editline.cpp:
```diff
@@ -393,7 +394,7 @@ void Editline::MoveCursor(CursorLocation from, CursorLocation to) {
   int toLine = GetLineIndexForLocation(to, editline_cursor_row);
   if (toLine != fromLine) {
-    fprintf(m_output_file,
+    fprintf(m_output_stream_sp->GetFile().GetStream(),
```
There are a lot of unguarded calls to `fprintf` and `fputs` in this file. Are there some guarantees about the state of the synchronized stream when these functions are executed? Or do we just have tons of unsynchronized writes to the output stream? :D

bulbazord reviewed (commented):
Overall looks fine, kind of terrified of the number of unguarded `fprintf` calls in Editline, but maybe it's fine?

labath commented on lldb/source/Interpreter/CommandInterpreter.cpp:
```diff
@@ -3108,13 +3108,11 @@ void CommandInterpreter::PrintCommandOutput(IOHandler &io_handler,
     llvm::StringRef line;
     std::tie(line, str) = str.split('\n');
     {
-      std::lock_guard<std::recursive_mutex> guard(io_handler.GetOutputMutex());
```
This is an example of an operation that used to be atomic, but now isn't.

labath reviewed (commented):
So, the problem I see here is that this is enforcing mutual exclusion at the level of most basic write operations. That's fine if what you want to achieve is to protect the internal state of the stream object. However, I'm not sure if that's your goal, since it doesn't e.g. prevent someone from printing something in the middle of a statusline update. It also may be unnecessary since I believe that both file descriptors and FILE* objects are synchronized internally (one by the OS, the other by the C library).

To achieve proper output synchronization, the exclusion needs to happen at a higher level (e.g. "a single statusline update", or "writing of one line of text", etc.). In theory, that can be done using the provided `GetMutex` accessor, but it's also very easy to forget doing that.

To better guarantee that, I am wondering if a slightly different pattern wouldn't be better. We could have one object (let's call it LockableStream here, just so it differs from what you have in this PR), which holds the real stream (and a mutex) as a *member*, and doesn't allow you to access it without locking it. Then, when you want write something, you call a `Lock` function or something (it could also be constructing a guard object and passing the LockableStream to its constructor), and this returns something that allows you to access the underlying stream. As long as you hold on to the returned object (which contains a lock_guard on the stream mutex internally), you are permitted to write to the stream. The destruction of the object can automatically flush the output stream.

WDYT?

JDevlieghere commented:
> it doesn't e.g. prevent someone from printing something in the middle of a statusline update.

Technically it does, because the statusline is printed as a single call to write, but I agree that's fragile and one could argue that works by accident rather than by design. I really like the idea of introducing an RAII object as it (1) makes the locking more explicit and (2) makes it easy to do the right thing and hard to do the wrong thing.

labath commented on lldb/source/Host/common/Editline.cpp:
```diff
@@ -700,12 +707,14 @@ unsigned char Editline::EndOrAddLineCommand(int ch) {
   }
   MoveCursor(CursorLocation::EditingCursor, CursorLocation::BlockEnd);
-  fprintf(m_output_file, "\n");
+  LockedStreamFile locked_stream = m_output_stream_sp->Lock();
```
And I suspect most/all of these callbacks are called from `GetLine` which means that they could reuse the `locked_stream` member instead of locking it once again.

labath commented on lldb/source/Commands/CommandObjectBreakpointCommand.cpp:
```diff
@@ -193,10 +193,12 @@ are no syntax errors may indicate that a function was declared but never called.
-      output_sp->Flush();
+    if (interactive) {
+      if (lldb::LockableStreamFileSP output_sp =
+              io_handler.GetOutputStreamFileSP()) {
+        LockedStreamFile locked_stream = output_sp->Lock();
+        locked_stream.PutCString(g_reader_instructions);
+      }
```
Optional, but I think it would be reasonable to write all of these "simple" lock calls as one-liners like this
```suggestion
              io_handler.GetOutputStreamFileSP())
        output_sp->Lock().PutCString(g_reader_instructions);
```

labath reviewed (commented):
Mostly yes. It's not as clean as I would have liked, though I suppose that is to be expected of an attempt to retrofit a significant change like this. The existence of Debugger::Get{Output,Error}File in particular is a very large hole in the RAII locking scheme. From the looks of things, none of the uses of these functions (GetErrorFile is unused already) is particularly load-bearing and they could be replaced by something else (but don't do that here, as the patch is big enough already).

The RAII thing doesn't make things much safer than the original implementation did because the Printf currently translate to a single `Write` call anyway, but they are more future-proof, as I can easily see someone "modernising" them into a `Format` call and then someone else "optimizing" our `Format` implementation to use the llvm raw_ostream method (which does *not* use a single write call). The current implementation will work even with those changes.

I do want to spend some time discussing the relationship between stdout and stderr. The current implementation uses a separate mutex for each stream, which is not *un*reasonable, but I can see at least two other possible philosophies:

1. Since both of the streams will be going to the same terminal most of the time, it might make sense to use a single mutex for both of them, as that's sort of the only way to ensure consistent output. (Imagine the situation of printing the "stderr" of a CLI command while the status line is being updated)
2. Alternatively, we could say that stderr is for immediate and exceptional output where it is more important to get the message to the user than it being formatted beautifully. In this world, we could keep "stderr" as a regular unlocked stream. In specific cases, where synchronizing the output is particularly important to us (like the beforementioned case of "stderr" of a CLI command) we could synchronize its output by holding the "stdout" mutex while writing it.

What do you think of that? I'm sort of leaning towards option two since stderr might be used in contexts where holding a mutex might not be completely safe (and the first option would make that even worse), and the current implementation with separate mutexes doesn't really guarantee "reasonable" stdout/err sequencing.

JDevlieghere commented:
> I do want to spend some time discussing the relationship between stdout and stderr. The current implementation uses a separate mutex for each stream, which is not _un_reasonable, but I can see at least two other possible philosophies:
> 
> 1. Since both of the streams will be going to the same terminal most of the time, it might make sense to use a single mutex for both of them, as that's sort of the only way to ensure consistent output. (Imagine the situation of printing the "stderr" of a CLI command while the status line is being updated)
> 2. Alternatively, we could say that stderr is for immediate and exceptional output where it is more important to get the message to the user than it being formatted beautifully. In this world, we could keep "stderr" as a regular unlocked stream. In specific cases, where synchronizing the output is particularly important to us (like the beforementioned case of "stderr" of a CLI command) we could synchronize its output by holding the "stdout" mutex while writing it.
> 
> What do you think of that? I'm sort of leaning towards option two since stderr might be used in contexts where holding a mutex might not be completely safe (and the first option would make that even worse), and the current implementation with separate mutexes doesn't really guarantee "reasonable" stdout/err sequencing.
  
I had a feeling this was going to come up... The tricky part is that, as you pointed out, most of the time the output is going to same terminal. But on the other hand, we totally support having those two streams go to totally separate files and it feels wrong to force them to be synchronized. How about a compromise between (1) and (2) is that we use the same mutex when the debugger is creating the StreamFile for stdout and stderr, but we use separate murexes when they're set through the setting (and the SB API)? 

Conceptually I'd also lean towards (2) but I worry about losing the benefits the current approach brings. Deciding whether "the output is particularly important" is a judgement call that we can't enforce at the API level. I like how the current implementation makes it easy to do the right thing and hard to do the wrong thing (or at least make you think about locking). If the default is to not lock the stdout (the alternative of always locking is equivalent to (1)) then I'm skeptical that we won't just end up with the same outcome that we have with two mutexes.

labath commented:
> > I do want to spend some time discussing the relationship between stdout and stderr. The current implementation uses a separate mutex for each stream, which is not _un_reasonable, but I can see at least two other possible philosophies:
> > 
> > 1. Since both of the streams will be going to the same terminal most of the time, it might make sense to use a single mutex for both of them, as that's sort of the only way to ensure consistent output. (Imagine the situation of printing the "stderr" of a CLI command while the status line is being updated)
> > 2. Alternatively, we could say that stderr is for immediate and exceptional output where it is more important to get the message to the user than it being formatted beautifully. In this world, we could keep "stderr" as a regular unlocked stream. In specific cases, where synchronizing the output is particularly important to us (like the beforementioned case of "stderr" of a CLI command) we could synchronize its output by holding the "stdout" mutex while writing it.
> > 
> > What do you think of that? I'm sort of leaning towards option two since stderr might be used in contexts where holding a mutex might not be completely safe (and the first option would make that even worse), and the current implementation with separate mutexes doesn't really guarantee "reasonable" stdout/err sequencing.
> 
> I had a feeling this was going to come up... The tricky part is that, as you pointed out, most of the time the output is going to same terminal. But on the other hand, we totally support having those two streams go to totally separate files and it feels wrong to force them to be synchronized. How about a compromise between (1) and (2) where we use the same mutex when the debugger is creating the StreamFile for stdout and stderr, but we use separate murexes when they're set through the setter (and the SB API)?

How would that work? We'd check  the value of the passed-in `FILE*` and choose a mutex based on whether it happens to be `stdout` ? I don't think I'd like that for several reasons:
- it doesn't handle the case where `stdout/err` is redirected externally (it still locks even though it shouldn't)
- it's doesn't let you create a terminal-like experience when embedding lldb-as-a-library (it doesn't lock even though it should)

> Conceptually I'd also lean towards (2) but I worry about losing the benefits the current approach brings. Deciding whether "the output is particularly important" is a judgement call that we can't enforce at the API level. I like how the current implementation makes it easy to do the right thing and hard to do the wrong thing (or at least make you think about locking). If the default is to not lock the stdout (the alternative of always locking is equivalent to (1)) then I'm skeptical that we won't just end up with the same outcome that we have with two mutexes.

But with two mutexes you still don't get stdout<=>stderr synchronization, which means that stderr output can corrupt things you're writing to stdout. What the stderr mutex buys you is stderr<=>stderr synchronization, which is... maybe nice, but maybe also not necessary?

FWIW, there is some precedent for unconditional stdout-stderr interaction: writes to `std::cerr` will automatically flush `std::cout` regardless of whether the two streams refer to the same terminal.  And the `iostream` library allows you to connect arbitrary two streams in this way. So, I think we *could* say we "tie" debuggers output and error streams in this way, even if they are completely independent, because they are "our" stdout and stderr.

JDevlieghere commented:
> How would that work? We'd check the value of the passed-in `FILE*` and choose a mutex based on whether it happens to be `stdout` ? I don't think I'd like that for several reasons:
> 
> * it doesn't handle the case where `stdout/err` is redirected externally (it still locks even though it shouldn't)
> * it's doesn't let you create a terminal-like experience when embedding lldb-as-a-library (it doesn't lock even though it should)

What I had in mind was much simpler than that: it would only apply when the debugger is the one creating the streams in its constructor. I was specifically referring to:

```
Debugger::Debugger(lldb::LogOutputCallback log_callback, void *baton)
    : UserID(g_unique_id++),
      Properties(std::make_shared<OptionValueProperties>()),
      m_input_file_sp(std::make_shared<NativeFile>(stdin, false)),
      m_output_stream_sp(std::make_shared<StreamFile>(stdout, false)),
      m_error_stream_sp(std::make_shared<StreamFile>(stderr, false)),
```

While not doing this when someone calls `SetOutputFile` or `SetErrorFile`. 

> > Conceptually I'd also lean towards (2) but I worry about losing the benefits the current approach brings. Deciding whether "the output is particularly important" is a judgement call that we can't enforce at the API level. I like how the current implementation makes it easy to do the right thing and hard to do the wrong thing (or at least make you think about locking). If the default is to not lock the stdout (the alternative of always locking is equivalent to (1)) then I'm skeptical that we won't just end up with the same outcome that we have with two mutexes.
> 
> But with two mutexes you still don't get stdout<=>stderr synchronization, which means that stderr output can corrupt things you're writing to stdout. What the stderr mutex buys you is stderr<=>stderr synchronization, which is... maybe nice, but maybe also not necessary?

Yes, I think we're saying the same thing here. 

> FWIW, there is some precedent for unconditional stdout-stderr interaction: writes to `std::cerr` will automatically flush `std::cout` regardless of whether the two streams refer to the same terminal. And the `iostream` library allows you to connect arbitrary two streams in this way. So, I think we _could_ say we "tie" debuggers output and error streams in this way, even if they are completely independent, because they are "our" stdout and stderr.

Fair enough. What I'm hearing is that you think it's important to synchronize the two somehow and I'd rather do it consistently rather than doing it only sometimes which makes it hard to know what the expectations both for us as the people writing that code and for our users. 

It should be fairly straightforward to support tying the two streams together. I'll see how that goes and update the PR.

labath commented:
> > How would that work? We'd check the value of the passed-in `FILE*` and choose a mutex based on whether it happens to be `stdout` ? I don't think I'd like that for several reasons:
> > 
> > * it doesn't handle the case where `stdout/err` is redirected externally (it still locks even though it shouldn't)
> > * it's doesn't let you create a terminal-like experience when embedding lldb-as-a-library (it doesn't lock even though it should)
> 
> What I had in mind was much simpler than that: it would only apply when the debugger is the one creating the streams in its constructor. I was specifically referring to:
> 
> ```
> Debugger::Debugger(lldb::LogOutputCallback log_callback, void *baton)
>     : UserID(g_unique_id++),
>       Properties(std::make_shared<OptionValueProperties>()),
>       m_input_file_sp(std::make_shared<NativeFile>(stdin, false)),
>       m_output_stream_sp(std::make_shared<StreamFile>(stdout, false)),
>       m_error_stream_sp(std::make_shared<StreamFile>(stderr, false)),
> ```
> 
> While not doing this when someone calls `SetOutputFile` or `SetErrorFile`.

I see. Well.. that's slightly cleaner, but it's currently a no-op because even our own driver calls [SetOutputFileHandle](https://github.com/llvm/llvm-project/blob/1199bbb396fb9554401ad5ae1816b6648bab76a9/lldb/tools/driver/Driver.cpp#L442). I guess we could change that, but I also don't particularly like the inaccessibility of this feature.

> 
> > > Conceptually I'd also lean towards (2) but I worry about losing the benefits the current approach brings. Deciding whether "the output is particularly important" is a judgement call that we can't enforce at the API level. I like how the current implementation makes it easy to do the right thing and hard to do the wrong thing (or at least make you think about locking). If the default is to not lock the stdout (the alternative of always locking is equivalent to (1)) then I'm skeptical that we won't just end up with the same outcome that we have with two mutexes.
> > 
> > 
> > But with two mutexes you still don't get stdout<=>stderr synchronization, which means that stderr output can corrupt things you're writing to stdout. What the stderr mutex buys you is stderr<=>stderr synchronization, which is... maybe nice, but maybe also not necessary?
> 
> Yes, I think we're saying the same thing here.

I guess so, but I think we're drawing different conclusions. My take on this is that if the only thing we get from locking stderr is the ability to mutually exclude two stderr accesses, then maybe we don't need to bother with locking stderr at all.

> > FWIW, there is some precedent for unconditional stdout-stderr interaction: writes to `std::cerr` will automatically flush `std::cout` regardless of whether the two streams refer to the same terminal. And the `iostream` library allows you to connect arbitrary two streams in this way. So, I think we _could_ say we "tie" debuggers output and error streams in this way, even if they are completely independent, because they are "our" stdout and stderr.
> 
> Fair enough. What I'm hearing is that you think it's important to synchronize the two somehow and I'd rather do it consistently rather than doing it only sometimes which makes it hard to know what the expectations both for us as the people writing that code and for our users.

Yes, I think that tying them *always* is better than tying them *sometimes*. However, I don't really have a clear preference between the other options flying around. The options being:
- tying them always
- not tying them at all
- not tying them and not locking stderr

If it sounds like I'm in favor of one of them, it's because I'm comparing its (dis)advantages relative to some other approach. However, I think all of these approaches have their (dis)advantages.

> It should be fairly straightforward to support tying the two streams together. I'll see how that goes and update the PR.

A random idea, I don't know if it's a good one: have one object (LockableStreamPair?) that holds the two streams and their mutex. One less argument to pass around?

JDevlieghere commented:
> > A random idea, I don't know if it's a good one: have one object (LockableStreamPair?) that holds the two streams and their mutex. One less argument to pass around?
> 
> Sounds good, that simplifies the signatures (which this PR is already touching) and keeps the mutex abstracted away.

I did implement this and I'm not happy with the result: even though you only pass one object, you still need to get the output or error stream out of it which ends up being more verbose. And because of the `IOHandler` and `Debugger::AdoptTopIOHandlerFilesIfInvalid` it needs to be a shared pointer. Here's what that looks like: https://github.com/llvm/llvm-project/commit/d62fdaec4cd322574e1ab78c8cacd1effe2c29c0. It compiles but some tests are failing. Probably something small but given I don't like the approach I haven't looked into it.

JDevlieghere commented on lldb/source/Commands/CommandObjectExpression.cpp:
```diff
@@ -500,19 +502,16 @@ bool CommandObjectExpression::EvaluateExpression(llvm::StringRef expr,
   io_handler.SetIsDone(true);
-  StreamFileSP output_sp = io_handler.GetOutputStreamFileSP();
-  StreamFileSP error_sp = io_handler.GetErrorStreamFileSP();
+  LockedStreamFile locked_output_stream =
+      io_handler.GetOutputStreamFileSP()->Lock();
+  LockedStreamFile locked_error_stream =
+      io_handler.GetErrorStreamFileSP()->Lock();
```
Fixed in ed32d85d31999756602a7d5c4647cb6771d8f857

labath commented:
> > > A random idea, I don't know if it's a good one: have one object (LockableStreamPair?) that holds the two streams and their mutex. One less argument to pass around?
> > 
> > 
> > Sounds good, that simplifies the signatures (which this PR is already touching) and keeps the mutex abstracted away.
> 
> I did implement this and I'm not happy with the result: even though you only pass one object, you still need to get the output or error stream out of it which ends up being more verbose. And because of the `IOHandler` and `Debugger::AdoptTopIOHandlerFilesIfInvalid` it needs to be a shared pointer. Here's what that looks like: [d62fdae](https://github.com/llvm/llvm-project/commit/d62fdaec4cd322574e1ab78c8cacd1effe2c29c0). It compiles but some tests are failing. Probably something small but given I don't like the approach I haven't looked into it.

That's okay. I thought it might not work out. Thanks for giving it a shot.

I feel bad about bringing this up after you did all the work, but there is one more thing going through my head. I now have a very awkward feeling about this because it seems like this is solving mostly the same problem that StreamAsynchronousIO (and Debugger::PrintAsync) was trying to solve. You mentioned that in the first version of the patch and said that it doesn't work in some cases, but it's not quite clear why is that the case. If the motivation is status line, then it sounds like it *might* be possible to plumb the updates through the async streams.

I can imagine that it might not work for some reason, but then I think it's important to know why, and come up with some sort of an story for which of the two systems should be used for a particular use case. Or, (I am dreaming, I know) actually deprecate/remove the async IO system.

JDevlieghere commented:
> That's okay. I thought it might not work out. Thanks for giving it a shot.
> 
> I feel bad about bringing this up after you did all the work, but there is one more thing going through my head. I now have a very awkward feeling about this because it seems like this is solving mostly the same problem that StreamAsynchronousIO (and Debugger::PrintAsync) was trying to solve. You mentioned that in the first version of the patch and said that it doesn't work in some cases, but it's not quite clear why is that the case. If the motivation is status line, then it sounds like it _might_ be possible to plumb the updates through the async streams.

The async streams buffer until they're flushed (or more commonly destroyed) and they delegate to the IOHandler instead of writing directly to the debugger's output or error stream. The problem is that while the IOHandler can protect the write, as is the case for `IOHandlerEditline` using the output mutex, nothing prevents anyone else from writing directly to the output stream, not using the asynchronous handler. If everyone uses the asynchronous streams, and every IOHandler would protect the write using the same mutex, then we could achieve the same. But that really files like working around the issue rather than addressing it (i.e. protecting the underlying streams), which is what this patch does. 

> I can imagine that it might not work for some reason, but then I think it's important to know why, and come up with some sort of an story for which of the two systems should be used for a particular use case. Or, (I am dreaming, I know) actually deprecate/remove the async IO system.

Yes, with this patch protecting the underlying streams, I don't think we _need_ the asynchronous streams and most uses could be converted to lock the stream. That said, there are still uses cases that benefit from the buffering, like running the expressions. As you pointed out, we don't want to lock the streams for the duration of running the expression. But maybe we should rename them to "BufferedOutputStream" but still remove the `PrintAsync` stuff that delegates to the IO handlers.

JDevlieghere commented on lldb/source/Host/common/Editline.cpp:
```diff
@@ -574,10 +581,10 @@ int Editline::GetCharacter(EditLineGetCharType *c) {
     // Read returns, immediately lock the mutex again and check if we were
     // interrupted.
-    m_output_mutex.unlock();
+    m_output_stream_sp->GetMutex().unlock();
```
Just hadn't gotten around to it. I've pushed a commit that adds this.

labath commented:
The buffering doesn't sound like a problem to me. Basically it means you have to use the stream like an RAII object, which is sort of what we have here anyway. We also have the same problem with not being able to get everything to use the RAII locker object from this patch (due to legacy reasons..)

That said, not needing to replicate the locking logic in every IOHandler would definitely be nice.
However, I'm not certain about the feasibility of that. In a way, I think that the PrintAsync approach is better because it lets the IOHandler know that some async printing is happening, and it can choose how to deal with it. For example, the ProcessIOHandler (forwarding stdout of the debugger proces) can end up in a state where the debugged process has written only a half a line. If async printing is plumbed through it, it will know that is happening, and it can choose to do something about (print an extra newline to delimit process output, or maybe erase the last output line, and then redraw it after the async output -- I don't know).

I'm not a mind reader nor a time traveller, but I have a feeling this is the reason why the async functionality is implemented the way it is. Maybe that's something we don't need? I think I could live without it. But if that is the case, I would like to see it dismantled, at least partially. For example, do you think it would be possible to change Debugger::PrintAsync to print the output directly (after locking the stream and such) -- without going through the IOHandlers and stuff?

JDevlieghere commented:
> That said, not needing to replicate the locking logic in every IOHandler would definitely be nice. However, I'm not certain about the feasibility of that. In a way, I think that the PrintAsync approach is better because it lets the IOHandler know that some async printing is happening, and it can choose how to deal with it. For example, the ProcessIOHandler (forwarding stdout of the debugger proces) can end up in a state where the debugged process has written only a half a line. If async printing is plumbed through it, it will know that is happening, and it can choose to do something about (print an extra newline to delimit process output, or maybe erase the last output line, and then redraw it after the async output -- I don't know).

Are you saying this is how it _could_ work? Looking at `IOHandlerProcessSTDIO`, it doesn't implement `PrintAsync` so it's just using the default implementation which prints to the debugger output stream.
 
> I'm not a mind reader nor a time traveller, but I have a feeling this is the reason why the async functionality is implemented the way it is. Maybe that's something we don't need? I think I could live without it. But if that is the case, I would like to see it dismantled, at least partially. For example, do you think it would be possible to change Debugger::PrintAsync to print the output directly (after locking the stream and such) -- without going through the IOHandlers and stuff?

I ran the test suite with this change:

```
diff --git a/lldb/source/Core/Debugger.cpp b/lldb/source/Core/Debugger.cpp
index 65f0502143bc..7e306fd7dcf4 100644
--- a/lldb/source/Core/Debugger.cpp
+++ b/lldb/source/Core/Debugger.cpp
@@ -1200,13 +1200,10 @@ bool Debugger::CheckTopIOHandlerTypes(IOHandler::Type top_type,
 }
 
 void Debugger::PrintAsync(const char *s, size_t len, bool is_stdout) {
-  bool printed = m_io_handler_stack.PrintAsync(s, len, is_stdout);
-  if (!printed) {
-    LockableStreamFileSP stream_sp =
-        is_stdout ? m_output_stream_sp : m_error_stream_sp;
-    LockedStreamFile locked_stream = stream_sp->Lock();
-    locked_stream.Write(s, len);
-  }
+  LockableStreamFileSP stream_sp =
+      is_stdout ? m_output_stream_sp : m_error_stream_sp;
+  LockedStreamFile locked_stream = stream_sp->Lock();
+  locked_stream.Write(s, len);
 }
 
 llvm::StringRef Debugger::GetTopIOHandlerControlSequence(char ch) {
```

I'm not sure how representative it is to rely on the test suite for something like this, but I got one failure (`TestBreakpointCallbackCommandSource.py`) because we don't redraw the prompt.

@labath After reading your last message I'm not sure where you want to take this. I understand the desire to not have two ways of doing things, but I don't necessarily see how they have to be tied together. I'm convinced that protecting the underlying streams is a net benefit. We were doing that in an ad-hoc way in the IOHandler previously and this moves it up a level. If that means we can improve other things (like the asynchronous IOHandler) then I'm happy to continue chipping away at this, but your earlier message sounds like you're proposing that as an alternative and I'm not sure I fully understand what you have in mind.

labath commented:
> > That said, not needing to replicate the locking logic in every IOHandler would definitely be nice. However, I'm not certain about the feasibility of that. In a way, I think that the PrintAsync approach is better because it lets the IOHandler know that some async printing is happening, and it can choose how to deal with it. For example, the ProcessIOHandler (forwarding stdout of the debugger proces) can end up in a state where the debugged process has written only a half a line. If async printing is plumbed through it, it will know that is happening, and it can choose to do something about (print an extra newline to delimit process output, or maybe erase the last output line, and then redraw it after the async output -- I don't know).
> 
> Are you saying this is how it _could_ work? Looking at `IOHandlerProcessSTDIO`, it doesn't implement `PrintAsync` so it's just using the default implementation which prints to the debugger output stream.

Correct. This is not how IOHandlerProcessSTDIO works right now (but [IOHandlerEditline](https://github.com/llvm/llvm-project/blob/e235fcb582eec5f58c905b66f96d0732d17b875e/lldb/source/Host/common/Editline.cpp#L1660) does). I'm using that as an example of something that would not be possible if one relied solely on the locking mechanism. The curses gui iohandler might be an even better example of that (although I'd much rather see it burn than improve it), as there the output might need to go to some particular curses window in order to avoid completely corrupting the curses window layout.

> 
> > I'm not a mind reader nor a time traveller, but I have a feeling this is the reason why the async functionality is implemented the way it is. Maybe that's something we don't need? I think I could live without it. But if that is the case, I would like to see it dismantled, at least partially. For example, do you think it would be possible to change Debugger::PrintAsync to print the output directly (after locking the stream and such) -- without going through the IOHandlers and stuff?
> 
> I ran the test suite with this change:
> 
> ```
> diff --git a/lldb/source/Core/Debugger.cpp b/lldb/source/Core/Debugger.cpp
> index 65f0502143bc..7e306fd7dcf4 100644
> --- a/lldb/source/Core/Debugger.cpp
> +++ b/lldb/source/Core/Debugger.cpp
> @@ -1200,13 +1200,10 @@ bool Debugger::CheckTopIOHandlerTypes(IOHandler::Type top_type,
>  }
>  
>  void Debugger::PrintAsync(const char *s, size_t len, bool is_stdout) {
> -  bool printed = m_io_handler_stack.PrintAsync(s, len, is_stdout);
> -  if (!printed) {
> -    LockableStreamFileSP stream_sp =
> -        is_stdout ? m_output_stream_sp : m_error_stream_sp;
> -    LockedStreamFile locked_stream = stream_sp->Lock();
> -    locked_stream.Write(s, len);
> -  }
> +  LockableStreamFileSP stream_sp =
> +      is_stdout ? m_output_stream_sp : m_error_stream_sp;
> +  LockedStreamFile locked_stream = stream_sp->Lock();
> +  locked_stream.Write(s, len);
>  }
>  
>  llvm::StringRef Debugger::GetTopIOHandlerControlSequence(char ch) {
> ```
> 
> I'm not sure how representative it is to rely on the test suite for something like this, but I got one failure (`TestBreakpointCallbackCommandSource.py`) because we don't redraw the prompt.

Thanks for trying it out. You're right, it probably isn't representative, but this test failure probably is a symptom of the problem -- this implementation would not trigger the prompt redrawing code I linked to above.

> @labath After reading your last message I'm not sure where you want to take this. I understand the desire to not have two ways of doing things, but I don't necessarily see how they have to be tied together.

I'm sorry about the confusion. I think that's a reflection of my ambiguity. I wouldn't really say I want to take this in a specific direction. I'm more of in the "asking questions" stage, where I'm turning the problem over in order to understand it.

Yesterday, I wasn't sure if we need the async stream mechanism. Today, I think we do (because of the prompt redrawing thingy). I think the two questions (async print and stream protection), but are separate in a way (the iohandlers still need, and the editline iohandler actually uses a mutex), but they are also kind of related in that they have to work together to achieve desired outcome ("nice" console output?) -- which may impact how they're implemented. 

At this point, I think the main question I am trying to answer is about a some guideline/rule about the usage of the various printing mechanism. Imagine I'm writing (or reviewing) a piece of code trying to print something. How do I know which method to use. If I can get a hold of a Debugger, I have the choice between GetOutputStreamSP, GetOutputFile and GetAsyncOutputStream. This patch doesn't really change that (which is why I think you say it's a net benefit), but it might make it seem like using GetOutputStreamSP (after locking it) is safe/okay -- which I think is the part that's bothering me.

So to try to propose an answer to this question: would it be correct to say that (even after this patch, as it is right now), one should approximately always use GetAsyncOutputStream for printing stuff out (at least in cases where one doesn't have the CommandReturnObject around), and that the other methods exist as implementation details of that (supporting locking in iohandlers, which is necessary to produce output correctly), or due legacy code we don't know how to get rid of?

JDevlieghere commented:
> Yesterday, I wasn't sure if we need the async stream mechanism. Today, I think we do (because of the prompt redrawing thingy). I think the two questions (async print and stream protection), but are separate in a way (the iohandlers still need, and the editline iohandler actually uses a mutex), but they are also kind of related in that they have to work together to achieve desired outcome ("nice" console output?) -- which may impact how they're implemented.

I think so too. When I started writing that message yesterday I was under the impression that we could get rid of the asynchronous streams altogether, but the prompt redrawing makes me doubt that. 

> At this point, I think the main question I am trying to answer is about a some guideline/rule about the usage of the various printing mechanism. Imagine I'm writing (or reviewing) a piece of code trying to print something. How do I know which method to use. If I can get a hold of a Debugger, I have the choice between GetOutputStreamSP, GetOutputFile and GetAsyncOutputStream. This patch doesn't really change that (which is why I think you say it's a net benefit), but it might make it seem like using GetOutputStreamSP (after locking it) is safe/okay -- which I think is the part that's bothering me.
>
> So to try to propose an answer to this question: would it be correct to say that (even after this patch, as it is right now), one should approximately always use GetAsyncOutputStream for printing stuff out (at least in cases where one doesn't have the CommandReturnObject around), and that the other methods exist as implementation details of that (supporting locking in iohandlers, which is necessary to produce output correctly), or due legacy code we don't know how to get rid of?

Yes. I now think that everyone should use the asynchronous streams by default. The lockable streams should only be used if the buffering poses a problem or you need the underlying stream for legacy reasons (e.g. the SB API). Let me try to confirm that assumption by converting all the uses that don't fall in the exception categories in a separate PR. If that works out, we can think about a way to enforce this at the API level (e.g. making the getters for the lockable stream private and relying on friends) and rebase this PR on top.

JDevlieghere commented:
Rebased on top of #127682

JDevlieghere commented:
As I had hoped, all the remaining uses of `LockableStreamFile` are now coming from the IOHandler.

labath reviewed (approved):
LGTM. Thanks for bearing with me.

JDevlieghere committed:
[lldb] Synchronize the debugger's output & error stream

This patch improves the synchronization of the debugger's output and
error streams using two new abstractions: LockableStreamFile and
LockedStreamFile.

 - LockableStreamFile is a wrapper around a StreamFile and a mutex.
   Client cannot use the StreamFile without calling Lock, which returns
   a LockedStreamFile.

 - LockedStreamFile is an RAII object that locks the stream for the
   duration of its existence. As long as you hold on to the returned
   object you are permitted to write to the stream. The destruction of
   the object automatically flush the output stream.
